docs/proposals/21-ipv6-singlestack-local.md

```diff
@@ -0,0 +1,287 @@
+Later on, the feature gate may also be used for safe-guarding maturing and enablement of a IPv6 single-stack implementation on cloud infrastructure.
+The feature gate is not supposed to be toggled back and forth.
+
+### `Shoot` API
+
+The `Shoot.spec.networking` section is extended as follows:
+
+```yaml
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+spec:
+  networking:
+    pods: 2001:db8:1::/48
+    nodes: 2001:db8:2::/48
+    services: 2001:db8:3::/48
+    # ...
+    ipFamilies:
+    - IPv6
+```
+
+`ipFamilies` is the central setting for specifying the IP families used for shoot networking.
+This field is inspired by the `Service.spec.ipFamilies` field in Kubernetes ([doc](https://kubernetes.io/docs/concepts/services-networking/dual-stack/#services)).
+The default value is `["IPv4"]` (IPv4 single-stack).
+If the `IPv6SingleStack` feature gate is enabled, `["IPv6"]` can be specified to switch to IPv6 single-stack.
+
+Later on, `["IPv4","IPv6"]` or `["IPv6","IPv4"]` can be supported for dual-stack networking.
+In dual-stack networking, the ordering of `ipFamilies` is needed for correctly configuring Kubernetes components, e.g., the API server's `--service-cluster-ip-range` flag controls which IP family is used to allocate the primary `clusterIP` of services.
+
+Instead of explicitly configuring the `ipFamilies` field, the used families could be determined implicitly from the corresponding CIDR fields.
+However, when using IPv6 single-stack networking on cloud infrastructure, users might want to use IPv6 prefixes assigned to them by the provider.
+Typically, these are not known by the user upfront and need to be allocated during cluster creation.
+In this case, users might not supply any of the CIDRs in their shoot specification.
+Hence, we need a central field that allows configuring the used IP families explicitly instead of implicitly.
+That's exactly the purpose of the `ipFamilies` field.
+
+If IPv6 single-stack networking is configured via the `ipFamilies` field, the existing `Shoot.spec.networking.{pods,services,nodes}` fields are used to specify the IPv6 CIDRs instead of the IPv4 CIDRs.
+
+All gardener components and extensions need to respect the `ipFamilies` field and handle it correctly, e.g., in API validation, defaulting, and configuring Shoot components.
+
+#### Future Dual-Stack Enhancements
+
+For supporting dual-stack networking setups in the future, the `Shoot` API must allow specifying CIDRs of both IP families.
+Similar to the [`Service` API](https://kubernetes.io/docs/concepts/services-networking/dual-stack/#services), the `Shoot` API may be extended with list equivalents of `Shoot.spec.networking.{pods,services,nodes}`.
+In this case, the existing CIDR fields may specify the respective CIDR of the primary IP family while the list fields contain CIDRs of both IP families.
+The primary IP family may be determined by the first element of the `ipFamilies` field.
```

Comment:
I am a bit confused now. If we use `pods`, `nodes` and `services` to specify the primary IP family why do we need the explicit `ipFamilies` field at all? It is certainly more verbose to have `ipFamilies`, but it also adds corner cases, e.g. what if I specify an IPv4 range in `pods` with `ipFamilies` set to `IPv6, IPv4`?

Reply:
I understand the intend to copy the behaviour of `Service.spec.ipFamilies` field in Kubernetes ([doc](https://kubernetes.io/docs/concepts/services-networking/dual-stack/#services)), but that one was already pretty bad in first place and won't stay that complex onece IPv6 is taking off.

Reply:
> I am a bit confused now. If we use `pods`, `nodes` and `services` to specify the primary IP family why do we need the explicit `ipFamilies` field at all?

In the end, the CIDRs API fields could be optional because users might want to use the provider-assigned ranges/prefixes (as it is already possible today for the `nodes` IPv4 CIDR).
In this case, they still need a central field to specify explicitly what kind of networking setup they want. That's what the `ipFamilies` field is for.
See [this section](https://github.com/stackitcloud/gardener/blob/ec25e096a264e44c0b5b6732ebc0f55024920035/docs/proposals/21-ipv6-singlestack-local.md?plain=1#L118-L123) of the document.

Also, having a dedicated and verbose `ipFamilies` field makes it simpler and more robust for all components (gardener and extensions) to determine which kind of networking setup needs to be realized.
If all components would try to parse the CIDRs to determine which kind of setup is used, this might lead to weird bugs and compatibility issues.

> but it also adds corner cases, e.g. what if I specify an IPv4 range in `pods` with `ipFamilies` set to `IPv6, IPv4`?

These corner cases should be prevented by API validation.
We added a sentence to the document accordingly.

Reply:
The keywords could be part of the providerConfig, as these configurations may vary over provider implementations and Gardener should not care where these CIDRs come from. We would leave it to the provider extensions to get these CIDRs and then Gardener will then use them.

However, since we are focusing on the local setup with IPv6 within Gardener we would prefer not to mention provider-specific things now.